firebase-perf/src/test/java/com/google/firebase/perf/application/FrameMetricsRecorderTest.java

```diff
@@ -0,0 +1,405 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.perf.application;
+
+import static com.google.common.truth.Truth.assertThat;
+import static org.mockito.ArgumentMatchers.argThat;
+import static org.mockito.ArgumentMatchers.nullable;
+import static org.mockito.Mockito.doCallRealMethod;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.inOrder;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.MockitoAnnotations.initMocks;
+
+import android.app.Activity;
+import android.util.SparseIntArray;
+import android.view.WindowManager;
+import androidx.core.app.FrameMetricsAggregator;
+import com.google.firebase.perf.FirebasePerformanceTestBase;
+import com.google.firebase.perf.metrics.FrameMetricsCalculator.PerfFrameMetrics;
+import com.google.firebase.perf.util.Optional;
+import java.util.List;
+import java.util.WeakHashMap;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.ArgumentCaptor;
+import org.mockito.InOrder;
+import org.mockito.Mock;
+import org.mockito.Spy;
+import org.robolectric.Robolectric;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.android.controller.ActivityController;
+
+/** Unit tests for {@link com.google.firebase.perf.application.FrameMetricsRecorder}. */
+@RunWith(RobolectricTestRunner.class)
+public class FrameMetricsRecorderTest extends FirebasePerformanceTestBase {
+  private Activity activity;
+  private FrameMetricsRecorder recorder;
+
+  @Mock private FrameMetricsAggregator fma;
+  @Mock private PerfFrameMetrics frameMetrics1;
+  @Mock private PerfFrameMetrics frameMetrics2;
+  @Mock private PerfFrameMetrics frameMetrics3;
+
+  @Spy private final WeakHashMap<Object, PerfFrameMetrics> subTraceMap = new WeakHashMap<>();
```

Comment:
This is a code-smell to me. Can we verify things through the public API without accessing internals? go/unit-testing-practices#public-apis

Reply:
Do you mean all the mocks or just the spy `subtraceMap`?

For `start()` it's hard to verify because it's a `void` function, not sure what's the best way to verify those.

Reply:
Following go/unit-testing-practices, now testing behaviors not individual methods, so this spy is removed.